### PR TITLE
[AIRFLOW-3274] Add run_as_user and fs_group options for Kubernetes

### PR DESCRIPTION
**Pull request by pgagnon**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3274
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

At this time it is not possible to add `run_as_user` or `fs_group` securityContext options to worker pods when using KubernetesExecutor. This makes it harder to use KubernetesExecutor on clusters with pod security policies which do not allow containers to run as root.

This PR adds `run_as_user` and `fs_group` options to Airflow in order to set these security context options on worker pods spawned by KubernetesExecutor.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

- `test_security_context` in `tests/contrib/minikube/test_kubernetes_pod_operator.py`

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

The PR adds configuration options documented with comments in the configuration template.

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: airflow/contrib/kubernetes/worker_configuration.py
Comment (skudriashev) [inline airflow/contrib/kubernetes/worker_configuration.py]: Please keep trailing comma.

File: airflow/contrib/kubernetes/worker_configuration.py
Comment (skudriashev) [inline airflow/contrib/kubernetes/worker_configuration.py]: Same here.

Comment (pgagnon) [reply airflow/contrib/kubernetes/worker_configuration.py]: @skudriashev Fixed.

File: airflow/contrib/kubernetes/worker_configuration.py
Comment (pgagnon) [reply airflow/contrib/kubernetes/worker_configuration.py]: Also fixed.

**Comment by pgagnon** [comment]
@ashb Apologies for the mention and double apologies if you aren't the right person for this. This PR has been sitting for a little while now and I'd just like to get a reviewer with write access interested 😊. This PR is important because it allows KubernetesExecutor to work on clusters with restrictive pod security policies in place.

**Comment by feng-tao** [comment]
@pgagnon , LGTM, we could merge the pr once you rebase with master,

**Comment by feng-tao** [comment]
cc @dimberman in case he has other comments.

File: airflow/contrib/kubernetes/worker_configuration.py
Comment (ashb) [inline airflow/contrib/kubernetes/worker_configuration.py]: Is it okay to have (for example) `fs_group: ""` which is I think what we would get if one of these options is provided in the config file but not the other.

Comment (pgagnon) [reply airflow/contrib/kubernetes/worker_configuration.py]: Yes, it would not cause any issues as if a field is empty the api-server ignores it.

Nevertheless I will refactor this block because this made me realize that the condition on line 238 will not scale well if more security context options are added in the future.

edit: In fact the condition on line 238 is useless so I will simply remove it.

**Comment by EamonKeane** [comment]
+1 

This adds the functionality needed on non-root pod operators whose entrypoints e.g. copy and chmod secret files (e.g. 0600 client-key.pem for postgres).

**Comment by pgagnon** [comment]
@dmateusp In this PR I would be modifying slightly the functionality you added in #4777, specifically with regards to how you are handling security context. If you have a chance could you please take a look  and let me know if you find any issues? Thanks in advance.

File: airflow/contrib/kubernetes/worker_configuration.py
Comment (dmateusp) [inline airflow/contrib/kubernetes/worker_configuration.py]: this will overwrite what the users set through the config, I'm sure it'll be a surprise

what about adding a comment in the config key saying:
```
# if using a git_ssh_key_secret_name use an fs_group
# that allows for the key to be read, e.g. 65533
```

And removing this line here

File: airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py
Comment (dmateusp) [inline airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py]: nit: doesn't it make more sense to just indent the code under a positive predicate instead of breaking the flow with a return, just for readability

```
if pod.security_context:
   ...
```

File: airflow/config_templates/default_airflow.cfg
Comment (dmateusp) [inline airflow/config_templates/default_airflow.cfg:730]: I'd add recommendations for git_sync usage as I mention in a comment below

File: tests/contrib/minikube/test_kubernetes_pod_operator.py
Comment (dmateusp) [inline tests/contrib/minikube/test_kubernetes_pod_operator.py]: What's this testing ? should you assert that the security context is indeed modified in the KubernetesPodOperator that is created ?

I don't think executing is enough

**Review by dmateusp — CHANGES_REQUESTED** [review]
Few minor changes/questions but this feature is useful +1

File: airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py
Comment (ashb) [reply airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py]: I prefer this style - rather than having the entire body of the function indented. (The rule I try to follow is to keep things as un-indented as possible.

File: airflow/contrib/kubernetes/worker_configuration.py
Comment (ashb) [reply airflow/contrib/kubernetes/worker_configuration.py]: Or `security_context.setdefault('fs_group', 65533)`?

File: tests/contrib/minikube/test_kubernetes_pod_operator.py
Comment (ashb) [reply tests/contrib/minikube/test_kubernetes_pod_operator.py]: Right now it's testing that it doesn't throw an exception, but I agree, more tests would be useful.

File: airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py
Comment (dmateusp) [reply airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py]: Ah I see! I guess it's a matter of opinion then, I tend to go with "not breaking the code flow", avoiding exceptions in the middle of the code, and impromptu returns so the code can be followed top to bottom

If that's the style followed in the Airflow project however no problems for me :)

File: tests/contrib/minikube/test_kubernetes_pod_operator.py
Comment (pgagnon) [reply tests/contrib/minikube/test_kubernetes_pod_operator.py]: I agree. I mostly mimicked the other similar tests for pod attributes but the whole test suite is lacking a bit in this regard.

File: airflow/contrib/kubernetes/worker_configuration.py
Comment (pgagnon) [reply airflow/contrib/kubernetes/worker_configuration.py]: I have addressed this comment by defaulting `fs_group` to `65533` if `git_ssh_key_secret_name` is present and `worker_fs_group` is not explicitely set to another value.

Also adding your suggestion to the config file template.

File: airflow/contrib/kubernetes/worker_configuration.py
Comment (pgagnon) [reply airflow/contrib/kubernetes/worker_configuration.py]: @ashb Actually I was wrong, this was causing integrations test to fails so I reverted to the previous behavior while taking into account the potential issue that you pointed out.

**Comment by pgagnon** [comment]
@ashb Rebased to master and fixed test. Ready for another review I think.

File: airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py
Comment (ashb) [reply airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py]: Style things often are  :D

File: airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py
Comment (ashb) [inline airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py]: This line used to be 

```
req['spec']['securityContext'] = pod.security_context
```

Now we pull it out of req.sec.securityContext - but I don't think that is ever set anywhere is it? Do we pull "base" values out of pod.security_context anywhere/anymore? Is req['spec']['securityContext']  ever set to anything?

```suggestion
        req['spec'].setdefault('securityContext', {})
```

I've lost all context on this PR and am can't follow through what comes from where.

Comment (pgagnon) [reply airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py]: @ashb I understand. 😃 The context is a bit harder to follow now since changes (f4253a29) were merged since this PR was submitted that touched the same area.

Previously the only option that was added from `pod.security_context` was `fsGroup = 65533`, if SSH private key support from f4253a29 was enabled, so we never actually pulled base values from `pod.security_context` (or anywhere else). This PR adds the ability to customize `fsGroup` and `runAsUser` while maintaining compatibility as much as possible with the changes introduced in the aforementioned commit.

Right now these are the only two security context options that are implemented in this PR, but it has been mentioned on slack that it might be beneficial to support the full range since some more locked down clusters might need them. I could add them but `fsGroup` and `runAsUser` are definitely the most common SC options so that's why I focused on them first.

I am committing your suggested change.

Comment (pgagnon) [reply airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py]: @ashb Any further questions or requested changes?

**Review by dimberman — DISMISSED** [review]
Overall this looks good to me, though I think we should modify the tests to create a non-root user for the deployment. Otherwise we might end up in cases where we break the executor for non-root users without knowing it.

**Comment by pgagnon** [comment]
@dimberman Agreed. I will implement that.

**Comment by pgagnon** [comment]
@dimberman Added more tests.

Also current test failure is a flake I think; it passed on my branch: https://travis-ci.org/pgagnon/incubator-airflow/builds/508313464

**Comment by pgagnon** [comment]
@dimberman @ashb From what I understand the window for 1.10.3 is closing soon. Do you think we can get this in? I am available to work on any further change requests.

File: airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py
Comment (ashb) [inline airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py]: `setdefault` modifies in place, so we don't need to/shouldn't assign the result.

Comment (ashb) [reply airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py]: Tiny nit: you've moved the location of this method in the file - could you put it back after `extract_tolerations` please?

**Review by ashb — COMMENTED** [review]
Two tiny changes then I'm happy.

@dimberman Final comments?

@pgagnon Yes, should be able to get this in to 1.10.3. I may start a `beta1` release without this, but I'll get it in the main release.

**Review by ashb — APPROVED** [review]

**Comment by mrunesson** [comment]
It would be good to consider the possibility to be able to set this per launched pod and also set the pod root filesystem to read only. See https://github.com/mrunesson/airflow/commit/ae2078d89e61fc420958dc337d57b993f2218eb3 for concept implementation.

**Comment by pgagnon** [comment]
> It would be good to consider the possibility to be able to set this per launched pod and also set the pod root filesystem to read only. See [mrunesson@ae2078d](https://github.com/mrunesson/airflow/commit/ae2078d89e61fc420958dc337d57b993f2218eb3) for concept implementation.

@mrunesson Thanks for the sample. I would implement this in a subsequent PR since this one is almost ready to be merged for 1.10.3.

**Comment by pgagnon** [comment]
@ashb Rebased on master to eliminate conflicts.

**Comment by ashb** [comment]
@pgagnon We just merged #4952 which adds a better way of testing  the request factory than launching pods. Could you update this PR  to use that style and then we can merge  this? Thanks

File: airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py
Comment (ashb) [inline airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py]: This is going to fail on Python2

Comment (pgagnon) [reply airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py]: My bad. I forgot we still supported Python 2.

**Review by ashb — APPROVED** [review]
✨

File: airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py
Comment (pgagnon) [reply airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py]: Fixed.

**Comment by ashb** [comment]
Test failures:

```
3) ERROR: test_make_pod_git_sync_ssh_without_known_hosts (tests.contrib.executors.test_kubernetes_executor.TestKubernetesWorkerConfiguration)
----------------------------------------------------------------------
   Traceback (most recent call last):
    tests/contrib/executors/test_kubernetes_executor.py line 421 in test_make_pod_git_sync_ssh_without_known_hosts
      self.assertEqual(65533, pod.security_context['fs_group'])
   KeyError: 'fs_group'
```

**Comment by pgagnon** [comment]
@ashb Failed again but I think it's a flake. https://travis-ci.org/apache/airflow/builds/516345620

**Comment by pgagnon** [comment]
@ashb All green. 🎉

**Comment by pgagnon** [comment]
@ashb Thanks for your patience and guidance. ☺️

**Pull request merged**
